onnxruntime/core/codegen/mti/nn/conv_ops.cc

```diff
@@ -0,0 +1,199 @@
+#include "core/codegen/mti/nn/conv_ops.h"
+
+#include "core/codegen/mti/math/matmul_ops.h"
+#include "core/codegen/mti/tensor/pad_ops.h"
+#include "core/codegen/mti/tensor/reshape_ops.h"
+#include "core/codegen/mti/tensor/transpose.h"
+
+namespace onnxruntime {
+namespace tvm_codegen {
+
+static tvm::Tensor PadTensor1D(const tvm::Tensor& input,
+                               const tvm::Array<tvm::Expr>& padding,
+                               size_t width_axis,
+                               const std::string& name) {
+  auto pad_left = padding[0];
+  auto pad_right = padding[1];
+
+  tvm::Array<tvm::Expr> pad_before(std::vector<tvm::Expr>(input->shape.size(), 0));
+  pad_before.Set(width_axis, pad_left);
+  tvm::Array<tvm::Expr> pad_after(std::vector<tvm::Expr>(input->shape.size(), 0));
+  pad_after.Set(width_axis, pad_right);
+
+  const int64_t* padding_w0 = tvm::as_const_int(pad_left);
+  const int64_t* padding_w1 = tvm::as_const_int(pad_right);
+
+  const bool do_pad = ((padding_w0 != nullptr && *padding_w0) ||
+                       (padding_w1 != nullptr && *padding_w1));
+
+  return do_pad ? Pad(input, pad_before, pad_after,
+                      0, "constant", name + "_input_padded")
+                : input;
+}
+
+tvm::Tensor Conv1D(const tvm::Tensor& input,
+                   const tvm::Tensor& filter,
+                   const tvm::Array<tvm::Expr>& out_shape,
+                   const tvm::Array<tvm::Expr>& stride,
+                   const tvm::Array<tvm::Expr>& padding,
+                   const std::string& name) {
+  size_t channel_axis = 1;
+  size_t width_axis = 2;
+
+  auto stride_width = stride[width_axis - 2];
+
+  auto input_padded = PadTensor1D(input, padding, width_axis, name);
+  auto rc = tvm::reduce_axis((tvm::Range(0, filter->shape[1])), "rc");
+  auto rx = tvm::reduce_axis((tvm::Range(0, filter->shape[2])), "rx");
+
+  return tvm::compute(
+      out_shape,
+      [&](const tvm::Array<tvm::Var>& output) {
+        tvm::Array<tvm::Expr> indices;
+        for (const tvm::Var& var : output) {
+          indices.push_back(var);
+        }
+        indices.Set(channel_axis, rc);
+        indices.Set(width_axis, output[width_axis] * stride_width + rx);
+
+        return tvm::sum(input_padded(indices) * filter({output[1], rc, rx}),
+                        {rc, rx});
+      },
+      name);
+}
+
+tvm::Tensor Conv2D(const tvm::Tensor& input,
+                   const tvm::Tensor& filter,
+                   const tvm::Array<tvm::Expr>& output_shape,
+                   const tvm::Array<tvm::Expr>& stride,
+                   const tvm::Array<tvm::Expr>& padding,
+                   const std::string& name) {
+  // Gemm Convolution
+  const int64_t* batch_size = tvm::as_const_int(input->shape[0]);
+  if (batch_size != nullptr && *batch_size == 1)
+    return Conv2D_gemm(input, filter, output_shape, stride, padding);
```

Comment:
I don't think we should put the conv2d_gemm here. It is efficient for some platform, but for others, like braisnlice, it can be scheduled in this way. You can keep this customized lowering in nuphar and register it during nuphar's graph iteration, but don't put it as generic conversion.

Reply:
I think there's value to share GEMM based conv2d implementation for ARM/X86?

---
In reply to: [294443494](https://github.com/microsoft/onnxruntime/pull/888#discussion_r294443494) [](ancestors = 294443494)

Reply:
but if i want to use this mti lib to convert Con2D to tvm, for any non-cpu target, i don't think i expect to get this Gemm based conv expression.
yes it could be shared between several cpu architecture, but I feel we need some clarify on this 'target-independent' mti lib, does this 'target-independent' means could works with different devices, or just different cpus?

---
In reply to: [297448544](https://github.com/microsoft/onnxruntime/pull/888#discussion_r297448544) [](ancestors = 297448544,294443494)

Reply:
Fixed by removing Conv2D_gemm from Conv2D

---
In reply to: [298400188](https://github.com/microsoft/onnxruntime/pull/888#discussion_r298400188) [](ancestors = 298400188,297448544,294443494)